### ISSUE TITLE
UDisks2 easily gets out of sync with mdadm

### ISSUE THREAD

**Issue by mvollmer** [storaged-project/udisks#613]
Using `mdadm` to make changes to a RAID device does not seem to trigger enough udev events always and the udev database gets out of sync.

One example is https://github.com/cockpit-project/cockpit/issues/9890,  Running (in my case)
```sh
mdadm --grow /dev/md127 --add /dev/sda --raid-devices=3 --level=0
```
only causes events on `/dev/sda`.

**Comment by vpodzime** [comment]
> only causes events on `/dev/sda`

Which should be enough, though, right? `/dev/sda` should become a RAID member and the particular RAID object(s) should be updated.

**Comment by mvollmer** [comment]
> > only causes events on `/dev/sda`
> 
> Which should be enough, though, right? `/dev/sda` should become a RAID member and the particular RAID object(s) should be updated.

Not completely.  Since only sda has a change event, the udev rules only run for sda and the UDISKS_MD_* properties of all other members and of the mdraid block device itself are not updated.  Also, UDisks2 doesn't pick up the new size of /dev/md127 after it has been grown, probably also because there was no change event for it.

**Comment by vpodzime** [comment]
> Since only sda has a change event, the udev rules only run for sda and the UDISKS_MD_* properties of all other members and of the mdraid block device itself are not updated. Also, UDisks2 doesn't pick up the new size of /dev/md127 after it has been grown, probably also because there was no change event for it.

Yes, these are the particular bugs. But in general a change event on one of the RAID members should be enough to trigger RAID object(s) update.

**Comment by mvollmer** [comment]
> But in general a change event on one of the RAID members should be enough to trigger RAID object(s) update.

Yes.  We could stop using udev properties and read everything from sysfs whenever there is a change event on any of the members or the raid device itself.

On the other hand, there is also /lib/udev/rules.d/63-md-raid-arrays.rules which suffers from the same problem, and it is shipped by mdadm itself.  After adding the third disk with the mdadm invocation above, /dev/md127 still has MD_DEVICES=2.  After "udevadm trigger", it changes to MD_DEVICES=3.

So this tells me that those udev change events are really missing, and that it's not a bug in UDisks2 to assume that they will happen.  (But we might still want to put a workaround into UDisks2, that's one of its roles after all.)
